File: ros2_ws/src/cpp_package/src/camera_node.cpp
#include <chrono>
#include <opencv2/opencv.hpp>
#include <sl/Camera.hpp>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/header.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "custom_interfaces/msg/image_input.hpp"
#include "cv_bridge/cv_bridge.h"

using namespace std::chrono_literals;


// ZED SDK Integration:
//      Uses sl::Mat for accessing ZED images.
//      Handles different views based on camera_side.
// ROS 2 Standard Practices:
//      Uses rclcpp::Node for ROS integration.
//      Publishes to a custom ROS message type.
// OpenCV Conversion:
//      Converts ZED's sl::Mat to OpenCV's cv::Mat for processing.
// Performance Logging:
//      Measures and logs preprocessing time.

class CameraNode : public rclcpp::Node {
public:
    CameraNode() : Node("camera_node") {
        RCLCPP_INFO(this->get_logger(), "Initializing Camera Node");

        // Declare and get parameters
        this->declare_parameter<std::string>("camera_side", "left");
        this->declare_parameter<std::vector<int>>("roi_dimensions", {0, 0, 100, 100});
        this->declare_parameter<int>("frame_rate", 30);

        camera_side_ = this->get_parameter("camera_side").as_string();
        roi_dimensions_ = this->get_parameter("roi_dimensions").as_integer_array();
        frame_rate_ = this->get_parameter("frame_rate").as_int();

        model_dimensions_ = cv::Size(640, 640);
        velocity_ = 0.0;
        shift_constant_ = (camera_side_ == "left") ? 0 : -1;

        // Publishers
        image_publisher_ = this->create_publisher<custom_interfaces::msg::ImageInput>(
            camera_side_ + std::string("_image_input"), 10);

        // Start ZED camera processing
        start_camera();
    }

private:
    void start_camera() {
        sl::Camera zed;
        sl::InitParameters init_params;
        init_params.camera_resolution = sl::RESOLUTION::HD1080;
        init_params.camera_fps = frame_rate_;

        if (camera_side_ == "left") {
            init_params.set_from_serial_number(left_camera_serial_number_);
        } else if (camera_side_ == "right") {
            init_params.set_from_serial_number(right_camera_serial_number_);
        }

        auto status = zed.open(init_params);
        if (status != sl::ERROR_CODE::SUCCESS) {
            RCLCPP_ERROR(this->get_logger(), "Failed to open ZED camera: %s", sl::toString(status).c_str());
            return;
        }

        sl::RuntimeParameters runtime_params;
        sl::Mat image_zed;

        while (rclcpp::ok()) {
            if (zed.grab(runtime_params) == sl::ERROR_CODE::SUCCESS) {
                zed.retrieve_image(image_zed, (camera_side_ == "left") ? sl::VIEW::LEFT_UNRECTIFIED : sl::VIEW::RIGHT_UNRECTIFIED);
                cv::Mat cv_image = slMatToCvMat(image_zed);
                preprocess_and_publish(cv_image);
                std::this_thread::sleep_for(std::chrono::milliseconds(1000 / frame_rate_));
            } else {
                RCLCPP_ERROR(this->get_logger(), "Failed to grab ZED frame");
            }
        }

        zed.close();
        RCLCPP_INFO(this->get_logger(), "Closing ZED camera");
    }

    void preprocess_and_publish(const cv::Mat &image) {
        auto start_time = std::chrono::high_resolution_clock::now();

        // Extract ROI
        int x1 = roi_dimensions_[0] + shift_constant_;
        int y1 = roi_dimensions_[1];
        int x2 = roi_dimensions_[2] + shift_constant_;
        int y2 = roi_dimensions_[3];
        cv::Mat roi_image = image(cv::Rect(cv::Point(x1, y1), cv::Point(x2, y2)));

        // Resize and convert to ROS message
        cv::Mat preprocessed_image;
        cv::resize(roi_image, preprocessed_image, model_dimensions_);

        auto raw_image_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", image).toImageMsg();
        auto preprocessed_image_msg = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", preprocessed_image).toImageMsg();

        // Publish custom message
        custom_interfaces::msg::ImageInput image_input;
        image_input.header.frame_id = "camera_frame";
        image_input.raw_image = *raw_image_msg;
        image_input.preprocessed_image = *preprocessed_image_msg;
        image_input.velocity = velocity_;

        image_publisher_->publish(image_input);

        auto end_time = std::chrono::high_resolution_clock::now();
        RCLCPP_INFO(this->get_logger(), "Preprocessing time: %.2f ms",
                    std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time).count());
    }

    cv::Mat slMatToCvMat(const sl::Mat &input) {
        int cv_type = (input.getDataType() == sl::MAT_TYPE::U8_C4) ? CV_8UC4 : CV_8UC3;
        return cv::Mat(input.getHeight(), input.getWidth(), cv_type, input.getPtr<sl::uchar1>(sl::MEM::CPU));
    }

    // Parameters
    std::string camera_side_;
    std::vector<int64_t> roi_dimensions_;
    int frame_rate_;

    // Camera settings
    int left_camera_serial_number_ = 26853647;
    int right_camera_serial_number_ = 0;
    double velocity_;
    int shift_constant_;
    cv::Size model_dimensions_;

    // ROS Publisher
    rclcpp::Publisher<custom_interfaces::msg::ImageInput>::SharedPtr image_publisher_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CameraNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: ros2_ws/src/python_package/python_package/camera_node.py
import time
import cv2
import pyzed.sl as sl

import rclpy
from rclpy.time import Time
from rclpy.node import Node
from rclpy.executors import MultiThreadedExecutor
from cv_bridge import CvBridge

from std_msgs.msg import Header
from sensor_msgs.msg import Image
from custom_interfaces.msg import ImageInput

from .scripts.utils import ModelInference
from .scripts.image_source import ImageSource

class CameraNode(Node):
    def __init__(self):
        super().__init__('camera_node')

        self.get_logger().info("Initializing Camera Node")
        
        self.declare_parameter('camera_side', 'left')
        self.declare_parameter('roi_dimensions', (0, 0, 100, 100))  
        self.declare_parameter('source_type', 'zed')
        self.camera_side = self.get_parameter('camera_side').get_parameter_value().string_value
        self.roi_dimensions = self.get_parameter('roi_dimensions').get_parameter_value().integer_array_value
        self.source_type = self.get_parameter('source_type').get_parameter_value().string_value
        
        # path = ""
        
        # # initialize image_source type
        # image = ImageSource(self.source_type, path)

        self.shift_constant = 0
        self.frame_rate = 1
        self.model_dimensions = (640, 640)
        self.velocity = 0.0
        self.bridge = CvBridge()
        self.left_camera_serial_number = 26853647
        self.right_camera_serial_number = 0

        self.input_image_publisher = self.create_publisher(ImageInput, f'{self.camera_side}_image_input', 10)
        
        self.publish_zed_frames()
    
    def publish_zed_frames(self):
        zed = sl.Camera()
        init = sl.InitParameters()
        init.camera_resolution = sl.RESOLUTION.HD1080
        init.camera_fps = self.frame_rate
        
        if self.camera_side == 'left':
            init.set_from_serial_number(self.left_camera_serial_number)
        elif self.camera_side == 'right':
            init.set_from_serial_number(self.right_camera_serial_number)
            self.shift_constant *= -1
        
        if not zed.is_opened():
            self.get_logger().info("Initializing Zed Camera")
        status = zed.open(init)
        if status != sl.ERROR_CODE.SUCCESS:
            self.get_logger().error(f"Failed to open ZED camera: {str(status)}")
            return
        
        runtime = sl.RuntimeParameters()
        image_zed = sl.Mat()
        
        while rclpy.ok():
            err = zed.grab(runtime)
            if err == sl.ERROR_CODE.SUCCESS:
                zed.retrieve_image(image_zed, sl.VIEW.LEFT_UNRECTIFIED)
                if self.camera_side == 'left':
                    zed.retrieve_image(image_zed, sl.VIEW.LEFT_UNRECTIFIED)
                else:
                    zed.retrieve_image(image_zed, sl.VIEW.RIGHT_UNRECTIFIED)
                
                cv_image = image_zed.get_data()  
                self.preprocess_image(cv_image)
                time.sleep(1.0 / self.frame_rate)
                
            else:
                self.get_logger().error("Failed to grab ZED camera frame: {str(err)}")

        zed.close()
        self.get_logger().info("Closing Zed Camera")
    
    def preprocess_image(self, image):
        tic = time.perf_counter_ns()
        roi_x1, roi_y1, roi_x2, roi_y2 = self.roi_dimensions
        shifted_x1 = roi_x1 + self.shift_constant
        shifted_x2 = roi_x2 + self.shift_constant

        preprocessed_img = image[int(roi_y1):int(roi_y2), int(shifted_x1):int(shifted_x2), :3] # replace w/ util
        preprocessed_img = cv2.resize(preprocessed_img, self.model_dimensions) # check if necessary?
        preprocessed_img_msg = self.bridge.cv2_to_imgmsg(preprocessed_img, encoding='rgb8')

        raw_image = image[:, :, :3]
        raw_img_msg = self.bridge.cv2_to_imgmsg(raw_image, encoding='rgb8')
        
        image_input = ImageInput()
        image_input.header = Header()
        image_input.header.frame_id = 'static_image' # fix
        image_input.raw_image = raw_img_msg
        image_input.preprocessed_image = preprocessed_img_msg
        image_input.velocity = self.velocity
        self.input_image_publisher.publish(image_input)
        toc = time.perf_counter_ns()
        self.get_logger().info(f"Preprocessing: {(toc-tic)/1e6} ms")
        time.sleep(1/self.frame_rate)

def main(args=None):
    rclpy.init(args=args)
    camera_node = CameraNode()
    executor = MultiThreadedExecutor(num_threads=1)
    executor.add_node(camera_node)
    
    try:
        executor.spin()
    except KeyboardInterrupt:
        print("Shutting down camera node")
    finally:
        executor.shutdown()
        camera_node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()
